onosproject/onos-lib-go
# Add logger grpc api

This PR is a working progress to implement a grpc api for the logger that we can control it from the CLI. 

To test, please follow the following steps:
0- Get this PR. You should replace the dep in onos-topo to point to this PR. 

1- Get this PR. https://github.com/onosproject/onos-topo/pull/92. This PR modifies the onos-topo cli and one of the go files to use the onos-lib-go logger. Update onos-cli dep to point to this new version of topo. 

2- cd onos-cli and make kind
3- Run make onos-topo-docker and then load it to kind cluster

4- onit create cluster --set onos-cli.enabled=true

5- Keep the logs in onos-topo running:  
`kubectl logs -n onos ONOS-TOPO-POD-NAME --follow `

6- We will modify the log level for store logger (i.e. the one which is defined in onos-topo PR using this PR)

7- Add a simulator : 
`onit add simulator -n sim1`

You should be able to see something like this since store logger using INFO level by default:
```
{"level":"INFO","ts":"2020-02-28T04:25:41.202Z","main":"main","caller":"onos-topo/onos-topo.go:100","msg":"Started NBI on [::]:5150"}
{"level":"INFO","ts":"2020-02-28T04:30:25.417Z","store":"store","caller":"device/service.go:184","msg":"Device is added"}
{"level":"INFO","ts":"2020-02-28T04:30:25.425Z","store":"store","caller":"device/service.go:184","msg":"Device is added"}
{"level":"INFO","ts":"2020-02-28T04:30:26.491Z","store":"store","caller":"device/service.go:151","msg":"Updated Device {}id:\"sim1\" revision:112 address:\"sim1:11161\" version:\"1.0.0\" timeout:<seconds:5 > credentials:<> tls:<plain:true > type:\"Devicesim\" protocols:<protocol:GNMI connectivityState:REACHABLE channelState:CONNECTED serviceState:AVAILABLE > "}
```

8- Change level to WARN level
`onit exec -- onos topo log set level store --level WARN `

9- Add another simulator. You shouldn't see the above output since we changed the log level to WARN. 

10- Change the log level to DEBUG: 
`onit exec -- onos topo log set level store  --level DEBUG`

11- Remove a simulator. You should be able to see something like this in the onos-topo log since DEBUG level prints INFO, WARN, ERROR levels (https://godoc.org/go.uber.org/zap/zapcore#Level)

```
{"level":"INFO","ts":"2020-02-28T04:34:45.018Z","store":"store","caller":"device/service.go:222","msg":"Info level device is removed"}
{"level":"DEBUG","ts":"2020-02-28T04:34:45.018Z","store":"store","caller":"device/service.go:223","msg":"Debug Level Device is removed"}
{"level":"WARN","ts":"2020-02-28T04:34:45.022Z","store":"store","caller":"device/service.go:190","msg":"device is removed"}
{"level":"WARN","ts":"2020-02-28T04:34:45.023Z","store":"store","caller":"device/service.go:190","msg":"device is removed"}
```
12- change log level to WARN and remove another simulator: 
`onit exec -- onos topo log set level store  --level WARN`

You should see just WARN messages like this: 
```
{"level":"WARN","ts":"2020-02-28T04:38:32.874Z","store":"store","caller":"device/service.go:190","msg":"device is removed"}
{"level":"WARN","ts":"2020-02-28T04:38:32.874Z","store":"store","caller":"device/service.go:190","msg":"device is removed"}
```
